In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
car_df=pd.read_csv("cardekho_dataset.csv",index_col=[0])
car_df.head()

car_name    brand     model  vehicle_age  km_driven seller_type  \
0    Maruti Alto   Maruti      Alto            9     120000  Individual   
1  Hyundai Grand  Hyundai     Grand            5      20000  Individual   
2    Hyundai i20  Hyundai       i20           11      60000  Individual   
3    Maruti Alto   Maruti      Alto            9      37000  Individual   
4  Ford Ecosport     Ford  Ecosport            6      30000      Dealer   

  fuel_type transmission_type  mileage  engine  max_power  seats  \
0    Petrol            Manual    19.70     796      46.30      5   
1    Petrol            Manual    18.90    1197      82.00      5   
2    Petrol            Manual    17.00    1197      80.00      5   
3    Petrol            Manual    20.92     998      67.10      5   
4    Diesel            Manual    22.77    1498      98.59      5   

   selling_price  
0         120000  
1         550000  
2         215000  
3         226000  
4         570000

#### perform feature engineering

In [3]:
car_df_copy=car_df.copy()

In [4]:
car_df_copy.head()

car_name    brand     model  vehicle_age  km_driven seller_type  \
0    Maruti Alto   Maruti      Alto            9     120000  Individual   
1  Hyundai Grand  Hyundai     Grand            5      20000  Individual   
2    Hyundai i20  Hyundai       i20           11      60000  Individual   
3    Maruti Alto   Maruti      Alto            9      37000  Individual   
4  Ford Ecosport     Ford  Ecosport            6      30000      Dealer   

  fuel_type transmission_type  mileage  engine  max_power  seats  \
0    Petrol            Manual    19.70     796      46.30      5   
1    Petrol            Manual    18.90    1197      82.00      5   
2    Petrol            Manual    17.00    1197      80.00      5   
3    Petrol            Manual    20.92     998      67.10      5   
4    Diesel            Manual    22.77    1498      98.59      5   

   selling_price  
0         120000  
1         550000  
2         215000  
3         226000  
4         570000

In [5]:
# check missing values
car_df_copy.isnull().any(axis=0)

car_name             False
brand                False
model                False
vehicle_age          False
km_driven            False
seller_type          False
fuel_type            False
transmission_type    False
mileage              False
engine               False
max_power            False
seats                False
selling_price        False
dtype: bool

### Apparently, there are no missing values in the dataset

In [6]:
# check column types
car_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15411 entries, 0 to 19543
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   car_name           15411 non-null  object 
 1   brand              15411 non-null  object 
 2   model              15411 non-null  object 
 3   vehicle_age        15411 non-null  int64  
 4   km_driven          15411 non-null  int64  
 5   seller_type        15411 non-null  object 
 6   fuel_type          15411 non-null  object 
 7   transmission_type  15411 non-null  object 
 8   mileage            15411 non-null  float64
 9   engine             15411 non-null  int64  
 10  max_power          15411 non-null  float64
 11  seats              15411 non-null  int64  
 12  selling_price      15411 non-null  int64  
dtypes: float64(2), int64(5), object(6)
memory usage: 1.6+ MB


In [7]:
car_df_copy["car_name"].value_counts()

car_name
Hyundai i20              906
Maruti Swift Dzire       890
Maruti Swift             781
Maruti Alto              778
Honda City               757
                        ... 
Mercedes-AMG C             1
Rolls-Royce Ghost          1
Maserati Quattroporte      1
Isuzu MUX                  1
Force Gurkha               1
Name: count, Length: 121, dtype: int64

In [8]:
# drop car_name column
car_df_copy.drop(labels="car_name",axis=1,inplace=True)

In [9]:
car_df_copy.head()

brand     model  vehicle_age  km_driven seller_type fuel_type  \
0   Maruti      Alto            9     120000  Individual    Petrol   
1  Hyundai     Grand            5      20000  Individual    Petrol   
2  Hyundai       i20           11      60000  Individual    Petrol   
3   Maruti      Alto            9      37000  Individual    Petrol   
4     Ford  Ecosport            6      30000      Dealer    Diesel   

  transmission_type  mileage  engine  max_power  seats  selling_price  
0            Manual    19.70     796      46.30      5         120000  
1            Manual    18.90    1197      82.00      5         550000  
2            Manual    17.00    1197      80.00      5         215000  
3            Manual    20.92     998      67.10      5         226000  
4            Manual    22.77    1498      98.59      5         570000

In [10]:
car_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15411 entries, 0 to 19543
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   brand              15411 non-null  object 
 1   model              15411 non-null  object 
 2   vehicle_age        15411 non-null  int64  
 3   km_driven          15411 non-null  int64  
 4   seller_type        15411 non-null  object 
 5   fuel_type          15411 non-null  object 
 6   transmission_type  15411 non-null  object 
 7   mileage            15411 non-null  float64
 8   engine             15411 non-null  int64  
 9   max_power          15411 non-null  float64
 10  seats              15411 non-null  int64  
 11  selling_price      15411 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 1.5+ MB


In [11]:
# brand column
car_df_copy["brand"].unique()

array(['Maruti', 'Hyundai', 'Ford', 'Renault', 'Mini', 'Mercedes-Benz',
       'Toyota', 'Volkswagen', 'Honda', 'Mahindra', 'Datsun', 'Tata',
       'Kia', 'BMW', 'Audi', 'Land Rover', 'Jaguar', 'MG', 'Isuzu',
       'Porsche', 'Skoda', 'Volvo', 'Lexus', 'Jeep', 'Maserati',
       'Bentley', 'Nissan', 'ISUZU', 'Ferrari', 'Mercedes-AMG',
       'Rolls-Royce', 'Force'], dtype=object)

In [12]:
# remove the brand column
car_df_copy.drop(labels="brand",axis=1,inplace=True)
car_df_copy.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      Alto            9     120000  Individual    Petrol            Manual   
1     Grand            5      20000  Individual    Petrol            Manual   
2       i20           11      60000  Individual    Petrol            Manual   
3      Alto            9      37000  Individual    Petrol            Manual   
4  Ecosport            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  selling_price  
0    19.70     796      46.30      5         120000  
1    18.90    1197      82.00      5         550000  
2    17.00    1197      80.00      5         215000  
3    20.92     998      67.10      5         226000  
4    22.77    1498      98.59      5         570000

In [13]:
car_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15411 entries, 0 to 19543
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   model              15411 non-null  object 
 1   vehicle_age        15411 non-null  int64  
 2   km_driven          15411 non-null  int64  
 3   seller_type        15411 non-null  object 
 4   fuel_type          15411 non-null  object 
 5   transmission_type  15411 non-null  object 
 6   mileage            15411 non-null  float64
 7   engine             15411 non-null  int64  
 8   max_power          15411 non-null  float64
 9   seats              15411 non-null  int64  
 10  selling_price      15411 non-null  int64  
dtypes: float64(2), int64(5), object(4)
memory usage: 1.4+ MB


In [14]:
# model column
car_df_copy["model"].unique()

array(['Alto', 'Grand', 'i20', 'Ecosport', 'Wagon R', 'i10', 'Venue',
       'Swift', 'Verna', 'Duster', 'Cooper', 'Ciaz', 'C-Class', 'Innova',
       'Baleno', 'Swift Dzire', 'Vento', 'Creta', 'City', 'Bolero',
       'Fortuner', 'KWID', 'Amaze', 'Santro', 'XUV500', 'KUV100', 'Ignis',
       'RediGO', 'Scorpio', 'Marazzo', 'Aspire', 'Figo', 'Vitara',
       'Tiago', 'Polo', 'Seltos', 'Celerio', 'GO', '5', 'CR-V',
       'Endeavour', 'KUV', 'Jazz', '3', 'A4', 'Tigor', 'Ertiga', 'Safari',
       'Thar', 'Hexa', 'Rover', 'Eeco', 'A6', 'E-Class', 'Q7', 'Z4', '6',
       'XF', 'X5', 'Hector', 'Civic', 'D-Max', 'Cayenne', 'X1', 'Rapid',
       'Freestyle', 'Superb', 'Nexon', 'XUV300', 'Dzire VXI', 'S90',
       'WR-V', 'XL6', 'Triber', 'ES', 'Wrangler', 'Camry', 'Elantra',
       'Yaris', 'GL-Class', '7', 'S-Presso', 'Dzire LXI', 'Aura', 'XC',
       'Ghibli', 'Continental', 'CR', 'Kicks', 'S-Class', 'Tucson',
       'Harrier', 'X3', 'Octavia', 'Compass', 'CLS', 'redi-GO', 'Glanza',
       

In [15]:
# separating categorical and numerical features
num_features=[feature for feature in car_df_copy.columns if car_df_copy[feature].dtype!='O']
cat_features=[feature for feature in car_df_copy.columns if car_df_copy[feature].dtype=='O']

print(num_features)
print(cat_features)

['vehicle_age', 'km_driven', 'mileage', 'engine', 'max_power', 'seats', 'selling_price']
['model', 'seller_type', 'fuel_type', 'transmission_type']


In [16]:
# separating independent and dependent features
X=car_df_copy.drop(labels="selling_price",axis=1)
y=car_df_copy["selling_price"]

In [17]:
X.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      Alto            9     120000  Individual    Petrol            Manual   
1     Grand            5      20000  Individual    Petrol            Manual   
2       i20           11      60000  Individual    Petrol            Manual   
3      Alto            9      37000  Individual    Petrol            Manual   
4  Ecosport            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  
0    19.70     796      46.30      5  
1    18.90    1197      82.00      5  
2    17.00    1197      80.00      5  
3    20.92     998      67.10      5  
4    22.77    1498      98.59      5

In [18]:
y

0         120000
1         550000
2         215000
3         226000
4         570000
          ...   
19537     250000
19540     925000
19541     425000
19542    1225000
19543    1200000
Name: selling_price, Length: 15411, dtype: int64

#### Feature encoding and scaling

In [19]:
# model column
len(X["model"].unique())

120

##### 'model' has 120 unique values, so performing OHE is not a very good idea, let's apply Label encoding since it will also give us the relationship between 'model' and 'selling_price'

In [20]:
from sklearn.preprocessing import LabelEncoder

le_encoder=LabelEncoder()
X["model"]=le_encoder.fit_transform(X["model"])

In [21]:
X.head()

model  vehicle_age  km_driven seller_type fuel_type transmission_type  \
0      7            9     120000  Individual    Petrol            Manual   
1     54            5      20000  Individual    Petrol            Manual   
2    118           11      60000  Individual    Petrol            Manual   
3      7            9      37000  Individual    Petrol            Manual   
4     38            6      30000      Dealer    Diesel            Manual   

   mileage  engine  max_power  seats  
0    19.70     796      46.30      5  
1    18.90    1197      82.00      5  
2    17.00    1197      80.00      5  
3    20.92     998      67.10      5  
4    22.77    1498      98.59      5

In [22]:
len(X["seller_type"].unique()),len(X["fuel_type"].unique()),len(X["transmission_type"].unique())

(3, 5, 2)

##### So for these columns, I can apply OHE, since very less categories, and rest on the numeric features apply standardscaling

In [23]:
# before that apply train test split
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)

In [24]:
from sklearn.preprocessing import OneHotEncoder,StandardScaler
from sklearn.compose import ColumnTransformer

ohe_columns=["seller_type","fuel_type","transmission_type"]
num_features=X.select_dtypes(exclude="object").columns

ohe_transformer=OneHotEncoder(drop="first")
scale_transformer=StandardScaler()

column_transform=ColumnTransformer(
    [
    ('ohe_transform',ohe_transformer,ohe_columns),
    ('scale_transformer',scale_transformer,num_features)
],remainder="passthrough")

In [25]:
# apply fit_transform on training data
X_train=column_transform.fit_transform(X_train)

# apply transform on test data
X_test=column_transform.transform(X_test)

In [26]:
X_train=pd.DataFrame(X_train,columns=column_transform.get_feature_names_out())
X_test=pd.DataFrame(X_test,columns=column_transform.get_feature_names_out())

# Now we'll do model training and model selection

In [34]:
from sklearn.linear_model import LinearRegression,Ridge,Lasso
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error

In [35]:
# a function to evaluate model
def evaluate_model(true_val,predicted_val):
    r2_val=r2_score(y_true=true_val,y_pred=predicted_val)
    mse=mean_squared_error(y_true=true_val,y_pred=predicted_val)
    mae=mean_absolute_error(y_true=true_val,y_pred=predicted_val)
    print(f"The R2 score is: {r2_val}")
    print(f"The mean square error: {mse}")
    print(f"The mean absolute error: {mae}")

In [36]:
# begin the model training
models={
    "Linear Regression":LinearRegression(n_jobs=-1),
    "Lasso":Lasso(),
    "Ridge":Ridge(),
    "SVR":SVR(),
    "KNN Regressor":KNeighborsRegressor(n_jobs=-1),
    "Decision Tree Regressor":DecisionTreeRegressor(),
    "Random Forest Regressor":RandomForestRegressor(n_jobs=-1),
    "AdaBoost Regressor":AdaBoostRegressor(),
    "Gradient Boosting Regressor":GradientBoostingRegressor()
}

for i in range(len(list(models))):
    model=list(models.values())[i]
    # train the model
    model.fit(X_train,y_train)

    # get predicted values for train data
    y_train_pred=model.predict(X_train)
    # get predicted values for test data
    y_test_pred=model.predict(X_test)

    print(f"Results for model: {list(models.keys())[i]}")
    # evaluate the model for training accuracy
    evaluate_model(true_val=y_train,predicted_val=y_train_pred)
    # evaluate the model for testing accuracy
    evaluate_model(true_val=y_test,predicted_val=y_test_pred)
    print("==============================================================")

Results for model: Linear Regression
The R2 score is: 0.6219860307551311
The mean square error: 304874315292.8461
The mean absolute error: 266675.10755425034
The R2 score is: 0.6524693637784766
The mean square error: 270286925822.75293
The mean absolute error: 284283.4459533829
Results for model: Lasso
The R2 score is: 0.6219860154615178
The mean square error: 304874327627.39
The mean absolute error: 266674.0558021801
The R2 score is: 0.6524701894197791
The mean square error: 270286283692.17163
The mean absolute error: 284283.80467042915
Results for model: Ridge
The R2 score is: 0.6219851714474167
The mean square error: 304875008338.2686
The mean absolute error: 266635.7643284719
The R2 score is: 0.6524827609405519
The mean square error: 270276506373.71832
The mean absolute error: 284241.57522491267
Results for model: SVR
The R2 score is: -0.05374480285789973
The mean square error: 849862045856.2803
The mean absolute error: 393329.8257361607
The R2 score is: -0.0683290674682584
The mea

### Perform hyperparameter tuning for Gradient-Boost regressor model

In [37]:
# hyperparameters for gradient-boost
loss=['squared_error', 'absolute_error', 'huber', 'quantile']
learning_rate=[0.01,0.1,1,10]
subsample=[0.5,0.1,1]
criterion=['friedman_mse','squared_error']
min_samples_split=[3,5,10]
min_samples_leaf=[3,5,10]
max_depth=[5,10]

gboost_params=dict(loss=loss,learning_rate=learning_rate,subsample=subsample,criterion=criterion,min_samples_split=min_samples_split,min_samples_leaf=min_samples_leaf,max_depth=max_depth)

In [38]:
# perform hyperparameter tuning
randomcv_models=[
    ("Gradient Boosting Regressor",GradientBoostingRegressor(),gboost_params)
]

In [39]:
from sklearn.model_selection import RandomizedSearchCV

model_params={}
for name,model,params in randomcv_models:
    random=RandomizedSearchCV(estimator=model,param_distributions=params,n_iter=20,n_jobs=-1,cv=3)
    random.fit(X_train,y_train)

    model_params[name]=random.best_params_

for model_name in model_params:
    print(f"---------------Best params for {model_name}------------------------")
    print(model_params[model_name])

e:\ml-bootcamp\venv\lib\site-packages\sklearn\model_selection\_search.py:1119: RuntimeWarning: overflow encountered in square
  (array - array_means[:, np.newaxis]) ** 2, axis=1, weights=weights


---------------Best params for Gradient Boosting Regressor------------------------
{'subsample': 1, 'min_samples_split': 3, 'min_samples_leaf': 3, 'max_depth': 10, 'loss': 'absolute_error', 'learning_rate': 0.1, 'criterion': 'friedman_mse'}


In [40]:
# now train the models with these params
models={
    "Gradient Boosting Regressor":GradientBoostingRegressor(subsample=1, min_samples_split=3, min_samples_leaf=3, max_depth=10, loss='absolute_error', learning_rate=0.1, criterion='friedman_mse')
}

for i in range(len(list(models))):
    model=list(models.values())[i]
    # train the model
    model.fit(X_train,y_train)

    # predict for train data
    y_train_pred=model.predict(X_train)
    # predict for test data
    y_test_pred=model.predict(X_test)

    model_name=list(models.keys())[i]
    print(f"Evaluation for {model_name}")
    print("------Train Data------")
    evaluate_model(y_train,y_train_pred)
    print("------Test Data------")
    evaluate_model(y_test,y_test_pred)
    print("================================")

Evaluation for Gradient Boosting Regressor
------Train Data------
The R2 score is: 0.9437969454507926
The mean square error: 45328662872.65825
The mean absolute error: 63690.684302209695
------Test Data------
The R2 score is: 0.9019235068273047
The mean square error: 76277574038.75832
The mean absolute error: 106218.28962021995
